cpp/bench/ann/conf/sift-128-euclidean.json

```diff
@@ -0,0 +1,1321 @@
+{
```

Comment:
By `configs` are you referring to parameter sets? The more I've been working through integrating these benchmarks, the more I'm realizing we should probably be specifying the parameters as ranges/lists, rather than individual hard-coded configs everywhere so that we can actually explore the qps/recall space and learn which parameters actually performed the best on which datasets. If we only test a very small set of parameters, we're not going to have all that great a representation of which ones worked the best. Even worse, it doesn't really give us a great reference point for comparison against other algorithms. 

The typical way I've seen to compare these algorithms is by allowing each algorithm to specify a range of reasonable parameter settings and then sweeping through all of them, which will yield a curve- qps drops as recall increases, and pulling off the pareto frontier. While looking around the [big-ann-benchmarks](https://github.com/harsha-simhadri/big-ann-benchmarks) documentation, I also noticed someone has created a repository that integrated the scripts almost directly from Erik Bernardsen's [ann-benchmarks](https://github.com/erikbern/ann-benchmarks/). I think we should consider doing the same (using their run/plot scripts, config infrastructure, etc...). 

Of course, this then begs the question- since all of the libraries we are comparing against have python apis, including our own, should we instead focus our efforts on benchmarking at the Python API? It would still allow us to make changes to the C++ code and instantly benchmark the results.

Reply:
- I am mostly referring to index building parameters. Index construction for 100M+ scale datasets can be long, so we probably want to limit to the most relevant configuration.
- For search parameters, I agree with you. Currently a parameter set sweeps trough 1 parameter (typically n_probbes). Having a way to sweep through a combination of parameters them would be nice.

> should we instead focus our efforts on benchmarking at the level of the Python API?

For small batch size the Python wrappers can have non negligible overhead. Therefore, I think there is a value in having benchmarks both for the C++ and Python APIs. But they do not have to provide the same level of convenience for configuring the test and reporting the results.

Reply:
> For small batch size the Python wrappers can have non negligible overhead. 

Yep, I agree- we should keep the c++ tests then. Benchmarking Dr the c++ layer also makes it much easier to benchmark the algorithms in the c++ layer. This is what Milvus does, for example. 

Maybe what I'll do is try to replicate a little bit more of the user experience from the Ann-benchmarks Python scripts, but only implement the options that make sense (for example, I don't think we need to worry about running docker containers, but it would be great it we could maybe specify the parameter ranges in a similar way.

It would be nice to use the plotting scripts too, and potentially the runner scripts. It could also be pretty easy to replicate their h5d file layout.

While all of the aforementioned things would be ideal the most ideal, I think we could probably merge this initial version after everyone's happy with it and then continue to iterate it on improving over the next couple releases. What do you think?